src/type/validate.js

```diff
@@ -593,6 +759,143 @@ function createInputObjectCircularRefsValidator(
   }
 }
 
+function createInputObjectDefaultValueCircularRefsValidator(
+  context: SchemaValidationContext,
+): (inputObj: GraphQLInputObjectType) => void {
+  // Modified copy of algorithm from 'src/validation/rules/NoFragmentCycles.js'.
+  // Tracks already visited types to maintain O(N) and to ensure that cycles
+  // are not redundantly reported.
+  const visitedFields = Object.create(null);
+
+  // Array of coordinates and default values used to produce meaningful errors.
+  const fieldPath = [];
+
+  // Position in the path
+  const fieldPathIndex = Object.create(null);
+
+  // This does a straight-forward DFS to find cycles.
+  // It does not terminate when a cycle was found but continues to explore
+  // the graph to find all possible cycles.
+  return function validateInputObjectDefaultValueCircularRefs(
+    inputObj: GraphQLInputObjectType,
+  ): void {
+    // Start with an empty object as a way to visit every field in this input
+    // object type and apply every default value.
+    return detectValueDefaultValueCycle(inputObj, {});
+  };
+
+  function detectValueDefaultValueCycle(
+    inputObj: GraphQLInputObjectType,
+    defaultValue: mixed,
+  ): void {
+    // If the value is a List, recursively check each entry for a cycle.
+    // Otherwise, only object values can contain a cycle.
+    if (isIterableObject(defaultValue)) {
+      for (const itemValue of defaultValue) {
+        detectValueDefaultValueCycle(inputObj, itemValue);
+      }
+      return;
+    } else if (!isObjectLike(defaultValue)) {
+      return;
+    }
+
+    // Check each defined field for a cycle.
+    for (const field of Object.values(inputObj.getFields())) {
+      const fieldType = getNamedType(field.type);
+
+      // Only input object type fields can result in a cycle.
+      if (!isInputObjectType(fieldType)) {
+        continue;
+      }
+
+      // If the provided value has this field defined, recursively check it for
+      // cycles instead of checking the field's default value.
+      if (hasOwnProperty(defaultValue, field.name)) {
+        detectValueDefaultValueCycle(fieldType, defaultValue[field.name]);
+        continue;
+      }
+
+      detectFieldDefaultValueCycle(inputObj, field, fieldType);
```

Comment:
I was confused when I read this at first because I figured lists could cause cycles, then noticed that fieldType was the _named_ type. Could we rename the variable to make this clearer, `fieldType` feels like an alias for  `field.type`
```suggestion
      const namedFieldType = getNamedType(field.type);

      // Only input object type fields can result in a cycle.
      if (!isInputObjectType(namedFieldType)) {
        continue;
      }

      // If the provided value has this field defined, recursively check it for
      // cycles instead of checking the field's default value.
      if (hasOwnProperty(defaultValue, field.name)) {
        detectValueDefaultValueCycle(namedFieldType, defaultValue[field.name]);
        continue;
      }

      detectFieldDefaultValueCycle(inputObj, field, namedFieldType);
```